### DISCUSSION TITLE
Add "Switch Network" modal event listener

### DISCUSSION THREAD

**Discussion by FrancoAguzzi** [rainbow-me/rainbowkit#1089] [Ideas]
Hey 👋🏼

I'm working on this dApp that allows Mainnet only. Once a user that is already connected changes his connected network via, e.g. Metamask, a "Switch Network" modal pops up asking the user to switch back to Mainnet. 

The problem is that user can simply close the modal and continue using the application while being connected to the wrong network. The behavior I want to implement would trigger a wallet disconnection if user closes the "Switch Network" modal without connecting to the correct network. 

For this, I'd need to know which action user took on "Switch Network" interaction, the problem is that today, we have no possibility of doing any programmatic interactions to this modal, we can't trigger it programmatically neither listen to its events (e.g. User closed modal).

First thing I'd like to know: is there any path I could follow to implement this behavior in the dApp I'm working on?

Secondly, I'd like to suggest the possibility of interacting with the built-in modals programmatically, or, at least listen to some user actioned events triggered when any modal interaction happens.

Please let me know your thoughts on these, thanks 🙏🏼

**Comment by FrancoAguzzi** [comment]
This is the modal I mentioned above:

<img width="674" alt="Captura de Tela 2023-03-17 às 18 35 27" src="https://user-images.githubusercontent.com/49823133/226057730-09231d9f-d623-41cf-b976-489473421fba.png">

**Comment by DanielSinclair** [comment]
@FrancoAguzzi You'll want to use the Wagmi `useNetwork` hook for this to react to the current selected chain/network:
```ts
const { chain } = useNetwork();
const { unsupported } = chain;
```
You can also use `useSwitchNetwork` to request MetaMask to switch the network. You would want to react to an unsupported network, then display an interface that prompts the user to change their network, and then attempt the change approval with `switchNetwork`.
```ts
const { switchNetwork } = useSwitchNetwork();
const handler = () => switchNetwork(mainnet);
```

**Reply by FrancoAguzzi** [reply]
@DanielSinclair I really appreciate your advice here, thanks for clarifying this workaround! 

This surely helps when creating a custom "Switch Network" modal is the case. 

Unfortunately, this is not the goal in our product for now, the best scenario would be to have this way of listening to modal events inside the scope of rainbowkit. Since we can't achieve this when directly interacting to rainbowkit, we chose to simply disconnect the user if he switches his current connected network while navigating through our app.

If you guys consider taking this next step on ChainModal development, let me know, I could help 😄

**Reply by DanielSinclair** [reply]
@FrancoAguzzi I might be missing something here. The user is alerted by RainbowKit that they have connected to an unsupported chain. They can toggle open this modal repeatedly until they've moved to a supported network.
<img width="377" alt="Screenshot 2023-03-22 at 11 35 01 PM" src="https://user-images.githubusercontent.com/4412473/227095641-0e53c360-399c-40b2-9a6f-810589e8310d.png">

The dApp also has the `chain` from `useNetwork()` and should gate the UI based on this unsupported status. The dApp can trigger a `switchNetwork` automatically or with user interaction to prompt the MetaMask user to switch back to a supported chain. Have a look at how Uniswap handles this; their UI enters a disabled state. This could be handled any number of ways by the dApp, like displaying a prompt or a login-like gate with a button to switch back to mainnet.
<img width="798" alt="Screenshot 2023-03-22 at 11 39 55 PM" src="https://user-images.githubusercontent.com/4412473/227096245-d627d63f-2ac1-4202-a269-73f4893dcd85.png">

**Reply by FrancoAguzzi** [reply]
This looks great. Thank you!
